Add table redirection support to SPI and Hive's redirection to Iceberg

lxynov opened the pull request:
### Add table redirection support to SPI 
Added these methods to `ConnectorMetadata`. Please review the Javadoc comments for intended engine behaviors. 
```java
/**
/**
 * Redirects to another table or view which may or may not be in the same catalog.
 * <p>
 * This method is called whenever the engine references a table or view name, including both accessing and creating one. Connectors are
 * expected to return {@link Optional#empty()} if the redirection doesn't happen. The engine calls this method in an iterative manner,
 * i.e., using the redirected name as the argument for the next call, until an {@link Optional#empty()} is returned. A
 * {@link StandardWarningCode#TABLE_REDIRECTION} warning is raised every time a redirection happens. Access control checks are performed
 * on the final table, i.e., the end of the redirection chain. Whether a table is a view or not is also determined by the final table.
 * E.g., suppose table a.b.c is redirected to x.y.z, it's a view if and only if x.y.z is a view.
 * </p>
 * <em>Caveats</em>:
 * <li>
 * If you choose to implement this method, you should consider implementing {@link #listTableColumnsStream}, {@link #getViewsStream},
 * {@link #listTablePrivilegesStream}, too. Please review their doc for expected behaviors on redirected tables. They are called to answer
 * information schema queries. You should also consider implementing {@link #redirectTableRename}, which is used to redirect rename targets.
 * </li>
 * <li>
 * Be careful if you implement redirection to another catalog. Different catalogs have different sets of roles. In other words, GRANT/REVOKE
 * queries may break due to the non-existence of certain roles in the new catalog. Different connectors may have different table properties,
 * system tables, hidden columns, type support, etc., too.
 * </li>
 */
default Optional<CatalogSchemaTableName> redirectTable(ConnectorSession session, SchemaTableName tableName)
{
    return Optional.empty();
}

/**
 * Redirects the target table or view name of a rename operation to another one which may or may not be in the same catalog.
 * <p>
 * This method is called before the engine tries to rename a table or view. Connectors are expected to return {@link Optional#empty()} if
 * the redirection doesn't happen. The engine calls this method and {@link #redirectTable} to redirect both source and target tables in an
 * iterative manner, until both methods return {@link Optional#empty()}. A {@link StandardWarningCode#TABLE_REDIRECTION} warning is raised
 * every time a redirection happens. Access control checks are performed on the final table,  i.e., the end of the redirection chain.
 * </p>
 * <em>Caveats</em>: See {@link #redirectTable}'s documentation for caveats of implementing table redirection.
 */
default Optional<CatalogSchemaTableName> redirectTableRename(ConnectorSession session, SchemaTableName sourceTableName, SchemaTableName targetTableName)
{
    return Optional.empty();
}
/**
 * Gets the metadata for all columns that match the specified table prefix. Returns a {@link ListTableColumnsResult}
 * with an Optional.empty() field value if a table is redirected.
 */
default Stream<ListTableColumnsResult> listTableColumnsStream(ConnectorSession session, SchemaTablePrefix prefix)
{
    return listTableColumns(session, prefix).entrySet().stream().map(entry -> new ListTableColumnsResult(entry.getKey(), Optional.of(entry.getValue())));
}

/**
 * Gets the definitions of views, possibly filtered by schema. Returns a {@link GetViewsResult}
 * with an Optional.empty() field value if a table is redirected.
 */
default Stream<GetViewsResult> getViewsStream(ConnectorSession session, Optional<String> schemaName)
{
    return getViews(session, schemaName).entrySet().stream().map(entry -> new GetViewsResult(entry.getKey(), Optional.of(entry.getValue())));
}

/**
 * List the table privileges granted to the specified grantee for the tables that have the specified prefix considering the selected session role.
 * Returns a {@link ListTablePrivilegesResult} with an Optional.empty() field value if a table is redirected.
 */
default Stream<ListTablePrivilegesResult> listTablePrivilegesStream(ConnectorSession session, SchemaTablePrefix prefix)
{
    return listTablePrivileges(session, prefix).stream().map(grantInfo -> new ListTablePrivilegesResult(grantInfo.getSchemaTableName(), Optional.of(grantInfo)));
}
```
### Add support to redirect from Hive to an Iceberg catalog
To enable this, we can either set Hive configs
```
hive.redirect-to-iceberg-enabled=true
hive.redirect-to-iceberg-catalog=iceberg
```
and keep session properties as default, or change session properties to
```
SET SESSION hive.redirect_to_iceberg_enabled = true;
SET SESSION hive.redirect_to_iceberg_catalog = 'iceberg';
```

---
Issue: #4442
Umbrella issue: #1324

findepi commented on presto-main/src/main/java/io/prestosql/execution/CommitTask.java:
```diff
@@ -37,7 +38,14 @@ public String getName()
+            AccessControl accessControl,
+            QueryStateMachine stateMachine,
+            List<Expression> parameters,
+            WarningCollector warningCollector)
```
Can i suggest separating the addition of the `WarningCollector` to the task interface, from other changes related to table redirection?

lxynov replied:
Sure, makes sense; updated.

lxynov commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -189,6 +192,8 @@
 
     private final ResolvedFunctionDecoder functionDecoder;
 
+    private static final int TABLE_REDIRECTION_MAX_TIMES = 100;
```
@electrum I re-included the max redirection enforcement to detect programming bugs that may lead to query execution hang. I thought that such bugs would be more common for table redirection than catalog redirection. Please let me know what you think.

electrum replied:
Let's name this `MAX_TABLE_REDIRECTIONS`

lxynov commented on presto-main/src/main/java/io/prestosql/execution/AddColumnTask.java:
```diff
@@ -72,6 +74,11 @@ public String getName()
+        Optional<QualifiedObjectName> redirectedTableName = metadata.redirectTable(session, tableName, !statement.isTableExists());
+        if (redirectedTableName.isPresent()) {
+            tableName = redirectedTableName.get();
+            warningCollector.add(new PrestoWarning(REDIRECTED_TABLE, "Table redirection happened"));
```
I firstly tried to include redirection details but later found that the `DefaultWarningCollector` stores warnings in a map keyed on `WarningCode`s, so a redirection warning's message can be overwritten by another one. So we'll need to implement another `WarningCollector` if we want to include details here. If we don't want to do so, it makes sense to move the warning to `MetadataManager`. @dain @electrum I'd leave it to you to make a call.

dain replied:
Interesting.  Let's leave this alone for now.

electrum commented on presto-main/src/main/java/io/prestosql/execution/CommentTask.java:
```diff
@@ -47,12 +50,24 @@ public String getName()
         if (statement.getType() == Comment.Type.TABLE) {
             QualifiedObjectName tableName = createQualifiedObjectName(session, statement, statement.getName());
+            Optional<QualifiedObjectName> redirectedTableName = metadata.redirectTable(session, tableName, true);
+            if (redirectedTableName.isPresent()) {
```
Move the warning inside of the `redirectTable()` method and include the source and destination in the warning message. This keeps the code simpler and we don't need to add warning collector everywhere.

Then this code can become
```java
tableName = metadata.redirectTable(session, tableName, true).orElse(tableName);
```
Or perhaps add a variant that returns either the redirection or the original.

electrum commented on presto-main/src/main/java/io/prestosql/execution/CreateTableTask.java:
```diff
@@ -154,14 +178,22 @@ public String explain(CreateTable statement, List<Expression> parameters)
                 if (!tableName.getCatalogName().equals(likeTableName.getCatalogName())) {
                     throw semanticException(NOT_SUPPORTED, statement, "LIKE table across catalogs is not supported");
                 }
-                TableHandle likeTable = metadata.getTableHandle(session, likeTableName)
```
It doesn't look like this code needs to change. I can't see what is different with regards to redirection.

electrum reviewed (changes requested):
The warning calls everywhere makes the code really cluttered. They need to go inside `MetadataManager`.

Another issue is how to handle metadata operations like `DESCRIBE`, `information_schema`, etc., and how those relate to the low level `ConnectorMetadata` calls.

This PR is good for review purposes of the SPI, but before we merge it, we need to add another commit with the Iceberg connector changes, and full test coverage of all the affected areas (i.e. make sure all the DDL commands are tested with redirection). We may need to introduce redirection into another connector like Memory for testing purposes (e.g., we could redirect from one memory catalog to another).

electrum commented:
We can introduce a replacement for `ConnectorMetadata.listTableColumns()`:
```java
Stream<ListColumnsResult> listColumns(ConnectorSession session, SchemaTablePrefix prefix);

class ListColumnsResult
{
    SchemaTableName tableName;
    Optional<CatalogSchemaTableName> redirection;
    List<ColumnMetadata> columns;
}
```
The `redirection` and `columns` properties are a union type (exactly one is set). For any results any results containing a redirection, the engine will recursively follow them and make the appropriate `listColumns()` calls.

The engine guarantees that the `Stream` is closed, in case connectors need that for resource management. The streaming part is an improvement over the existing API as it means `LIMIT` queries could execute faster.

electrum commented:
We need a similar change for `getViews()`.

electrum commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -587,31 +594,43 @@ public ColumnMetadata getColumnMetadata(Session session, TableHandle tableHandle
 
                 ConnectorSession connectorSession = session.toConnectorSession(catalogName);
-                for (Entry<SchemaTableName, List<ColumnMetadata>> entry : metadata.listTableColumns(connectorSession, tablePrefix).entrySet()) {
+                for (Entry<SchemaTableName, Optional<List<ColumnMetadata>>> entry : metadata.listTableColumnsWithRedirection(connectorSession, tablePrefix).entrySet()) {
```
Using `forEach` here allows naming the key/value pairs:
```java
metadata.listTableColumnsWithRedirection(connectorSession, tablePrefix).forEach((name, columns) -> {
    ...
});

electrum commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -587,31 +594,43 @@ public ColumnMetadata getColumnMetadata(Session session, TableHandle tableHandle
+            // handle redirections
+            for (QualifiedObjectName redirectedTable : redirectedTables) {
+                QualifiedObjectName finalTable = redirectTable(session, redirectedTable, WarningCollector.NOOP);
+                getTableHandle(session, finalTable).ifPresent(handle -> tableColumns.put(redirectedTable, getTableMetadata(session, handle).getColumns()));
```
This is a bit long, I'd wrap here
```java
getTableHandle(session, finalTable).ifPresent(handle ->
        tableColumns.put(redirectedTable, getTableMetadata(session, handle).getColumns()));

electrum commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -587,31 +594,43 @@ public ColumnMetadata getColumnMetadata(Session session, TableHandle tableHandle
+            }
+
+            // if table and view names overlap, the view wins
+            for (Entry<QualifiedObjectName, ConnectorViewDefinition> entry : getViews(session, prefix).entrySet()) {
```
```java
getViews(session, prefix).forEach((viewName, columns) -> {
    ...
});

electrum commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -962,7 +981,11 @@ public void finishDelete(Session session, TableHandle tableHandle, Collection<Sl
         if (objectName.isPresent()) {
-            return getView(session, objectName.get())
+            // redirectTable assumes a non-empty table name
+            if (objectName.get().getCatalogName().isEmpty() || objectName.get().getSchemaName().isEmpty() || objectName.get().getObjectName().isEmpty()) {
```
This shouldn't be needed, since `asQualifiedObjectName()` only returns a present optional if the name is fully qualified.

lxynov replied:
The comment `// redirectTable assumes a non-empty table name` is a little bit misleading. Here's actually to handle empty strings. Example query:
```SQL
SELECT * FROM hive.information_schema.views WHERE table_name = '';
```
Let me change the comment to
```
// View cannot exist
```
to be consistent with the one in `MetadataManager::getView`.

electrum commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -997,26 +1020,31 @@ public void finishDelete(Session session, TableHandle tableHandle, Collection<Sl
+                        if (entry.getValue().isPresent()) {
+                            views.put(viewName, entry.getValue().get());
+                        }
+                        redirectedViews.add(viewName);
```
Should this be in an else block?

electrum commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -1463,12 +1491,34 @@ public void revokeTablePrivileges(Session session, QualifiedObjectName tableName
+    {
+        ImmutableList.Builder<GrantInfo> grantInfos = ImmutableList.builder();
+        ImmutableList.Builder<QualifiedObjectName> redirectedTables = ImmutableList.builder();
+        for (Entry<SchemaTableName, Optional<List<GrantInfo>>> entry : listTablePrivilegesWithRedirection(session, prefix).entrySet()) {
```
```java
listTablePrivilegesWithRedirection(session, prefix).forEach((name, grantInfo) -> {

electrum commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -1463,12 +1491,34 @@ public void revokeTablePrivileges(Session session, QualifiedObjectName tableName
+        }
+        for (QualifiedObjectName redirectedTable : redirectedTables.build()) {
+            QualifiedObjectName finalTable = redirectTable(session, redirectedTable, WarningCollector.NOOP);
+            listTablePrivilegesWithRedirection(session, finalTable.asQualifiedTablePrefix()).values().forEach(tableGrantInfos -> tableGrantInfos.ifPresent(grantInfos::addAll));
```
This is a bit long, consider wrapping
```java
listTablePrivilegesWithRedirection(session, finalTable.asQualifiedTablePrefix()).values()
        .forEach(tableGrantInfos -> tableGrantInfos.ifPresent(grantInfos::addAll));

electrum commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -2064,6 +2114,84 @@ public AnalyzePropertyManager getAnalyzePropertyManager()
+        Set<QualifiedObjectName> visitedTableNames = new HashSet<>();
+        visitedTableNames.add(tableName);
+        int count = TABLE_REDIRECTION_MAX_TIMES;
+        while (count-- > 0) {
```
We avoid using increment/decrement in an expression. This can be written as a traditional for loop
```java
for (int count = 0; count < MAX_TABLE_REDIRECTIONS; count++) {

electrum commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -2064,6 +2114,84 @@ public AnalyzePropertyManager getAnalyzePropertyManager()
+            if (redirectedTableName.isEmpty()) {
+                return tableName;
+            }
+            warningCollector.add(new PrestoWarning(REDIRECTED_TABLE, format("Reference to table or view %s is redirected to %s", tableName, redirectedTableName.get())));
```
Maybe shorten to
```java
"Table or view '%s' redirected to '%s'"

electrum commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -2064,6 +2114,84 @@ public AnalyzePropertyManager getAnalyzePropertyManager()
+            warningCollector.add(new PrestoWarning(REDIRECTED_TABLE, format("Reference to table or view %s is redirected to %s", tableName, redirectedTableName.get())));
+            tableName = redirectedTableName.get();
+            if (!visitedTableNames.add(tableName)) {
+                throw new PrestoException(GENERIC_INTERNAL_ERROR, "Table redirection forms a loop");
```
Let's add an error code for this and include the original name
```java
throw new PrestoException(TABLE_REDIRECTION_LOOP, "Redirection for table forms a loop: " + originalTableName);

findepi replied:
Maybe include the whole cycle?
Otherwise with a redirect like `A -> B -> C -> D -> E -> D` it will be hard to identity D and E loop

lxynov replied:
Added the whole cycle

electrum commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -2064,6 +2114,84 @@ public AnalyzePropertyManager getAnalyzePropertyManager()
+                throw new PrestoException(GENERIC_INTERNAL_ERROR, "Table redirection forms a loop");
+            }
+        }
+        throw new PrestoException(GENERIC_INTERNAL_ERROR, format("Table redirection happened more than %d times", TABLE_REDIRECTION_MAX_TIMES));
```
Add error code and include original name
```java
throw new PrestoException(TABLE_REDIRECTION_LIMIT, format("Too many redirections (%s) for table: %s", MAX_TABLE_REDIRECTIONS, originalTableName));

findepi replied:
If this exception is raised, administrator should have a way to understand what's wrong.
Consider including the redirects list in the message.

electrum commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -2064,6 +2114,84 @@ public AnalyzePropertyManager getAnalyzePropertyManager()
+    }
+
+    @Override
+    public RenameRedirection redirectTableRename(Session session, QualifiedObjectName sourceTableName, QualifiedObjectName targetTableName, WarningCollector warningCollector)
```
Same comments as for `redirectTable()`

electrum commented on presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java:
```diff
@@ -364,7 +364,7 @@ protected Scope visitUse(Use node, Optional<Scope> scope)
         protected Scope visitInsert(Insert insert, Optional<Scope> scope)
         {
-            QualifiedObjectName targetTable = createQualifiedObjectName(session, insert, insert.getTarget());
+            QualifiedObjectName targetTable = metadata.redirectTable(session, createQualifiedObjectName(session, insert, insert.getTarget()), warningCollector);
```
We might be able to add a helper method in this class
```java
QualifiedObjectName targetTable = getRedirectedTable(insert, insert.getTarget());

private QualifiedObjectName getRedirectedTable(Node node, QualifiedName name) { ... }

lxynov replied:
Hmm this helper method won't help reduce the amount of code so I didn't use it in the updated PR. Please let me know if you think it's still better to use it.

electrum commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -223,6 +226,14 @@ default ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTabl
+     */
+    default Map<SchemaTableName, Optional<List<ColumnMetadata>>> listTableColumnsWithRedirection(ConnectorSession session, SchemaTablePrefix prefix)
+    {
+        return listTableColumns(session, prefix).entrySet().stream().collect(toUnmodifiableMap(Map.Entry::getKey, entry -> Optional.of(entry.getValue())));
```
Wrap the `collect()` call
```java
return listTableColumns(session, prefix).entrySet().stream()
        .collect(toUnmodifiableMap(Map.Entry::getKey, entry -> Optional.of(entry.getValue())));

electrum commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -223,6 +226,14 @@ default ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTabl
     }
 
+    /**
+     * Returns a (table name, Optional.empty()) entry if the table is redirected. See {@link #redirectTable}'s doc for details.
```
Use the `{@code }` tag
```java
/**
 * Returns a {@code (tableName, Optional.empty())} entry if the table is redirected. See {@link #redirectTable} for details.
 */
```
Same for others below

electrum reviewed (commented):
The engine changes look good. I'm still reviewing the Iceberg commit

electrum commented on presto-spi/src/main/java/io/prestosql/spi/connector/StandardWarningCode.java:
```diff
@@ -22,6 +22,7 @@
     TOO_MANY_STAGES(0x0000_0001),
     REDUNDANT_ORDER_BY(0x0000_0002),
     DEPRECATED_FUNCTION(0x0000_0003),
+    REDIRECTED_TABLE(0x0000_0004),
```
`TABLE_REDIRECTION` sounds better to me, since it makes the "redirection" the noun rather than the "table". Thoughts?

lxynov replied:
Sounds better to me, too

electrum commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -1021,4 +1050,54 @@ default MaterializedViewFreshness getMaterializedViewFreshness(ConnectorSession
+     * This method is called whenever the engine references a table or view name, including both accessing and creating one. Connectors are
+     * expected to return {@link Optional#empty()} if the redirection doesn't happen. The engine calls this method in an iterative manner,
+     * i.e., using the redirected name as the argument for the next call, until an {@link Optional#empty()} is returned. If a redirection
+     * does happen, the engine will access the redirected table or view name instead of the original one and raise a REDIRECTED_TABLE warning.
```
Use a `@link` for the warning code. Same for comment below

lxynov commented on presto-main/src/main/java/io/prestosql/execution/AddColumnTask.java:
```diff
@@ -71,7 +71,7 @@ public String getName()
     {
         Session session = stateMachine.getSession();
-        QualifiedObjectName tableName = createQualifiedObjectName(session, statement, statement.getName());
+        QualifiedObjectName tableName = metadata.redirectTable(session, createQualifiedObjectName(session, statement, statement.getName()), warningCollector);
```
Agreed. But making separate calls makes `tableName` not "effectively final", so it can't be referenced in lambdas. In this case, we have to make changes around lambda calls that seem irrelevant, e.g, wiping out lambda calls or declaring a new final reference. So it's like a trade-off between readabilities of code and commit. What do you think?

findepi commented on presto-main/src/main/java/io/prestosql/metadata/Metadata.java:
```diff
@@ -571,4 +573,62 @@ default ResolvedFunction getCoercion(Type fromType, Type toType)
+     */
+    default QualifiedObjectName redirectTable(Session session, QualifiedObjectName tableName, WarningCollector warningCollector)
+    {
+        return tableName;
```
There are 2 implementations of this interface (1 non-test), so no need for a default implementation.

findepi commented on presto-main/src/main/java/io/prestosql/metadata/Metadata.java:
```diff
@@ -571,4 +573,62 @@ default ResolvedFunction getCoercion(Type fromType, Type toType)
+     */
+    default RenameRedirection redirectTableRename(Session session, QualifiedObjectName sourceTableName, QualifiedObjectName targetTableName, WarningCollector warningCollector)
+    {
+        return new RenameRedirection(sourceTableName, targetTableName);
```
This gets overridden, so please remove the default implementation, can be misleading.

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -1021,4 +1050,54 @@ default MaterializedViewFreshness getMaterializedViewFreshness(ConnectorSession
     }
+
+    /**
+     * Redirects to another table or view which may or may not in the same catalog.
```
"may or may not" -> may or may not be

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -223,6 +226,14 @@ default ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTabl
+     */
+    default Map<SchemaTableName, Optional<List<ColumnMetadata>>> listTableColumnsWithRedirection(ConnectorSession session, SchemaTablePrefix prefix)
+    {
+        return listTableColumns(session, prefix).entrySet().stream().collect(toUnmodifiableMap(Map.Entry::getKey, entry -> Optional.of(entry.getValue())));
```
Shouldn't the default implementation actually call `redirectTable`?

Something like
```
return listTableColumns(session, prefix).entrySet().stream()
        .map(entry -> {
            if (redirectTable(session, entry.getKey()).isPresent()) {
                return new SimpleEntry<SchemaTableName, Optional<List<ColumnMetadata>>>(entry.getKey(), Optional.empty());
            }
            return new SimpleEntry<>(entry.getKey(), Optional.of(entry.getValue()));
        })
        .collect(toUnmodifiableMap(Entry::getKey, Entry::getValue));
```

lxynov replied:
This makes sense. But I think there's a trade-off here:
1. Let the default implementation call `redirectTable`. So it's fine even if a connector implements `redirectTable` and not `listTableColumnsWithRedirection`. However, this makes other connectors less efficient, e.g. those which don't implement table redirection APIs
2. Don't let the default implementation call `redirectTable` and make connector authors responsible for also implementing `listTableColumnsWithRedirection` when implementing `redirectTable`. The pro and con are the contrary.

What do you folks think? @findepi @electrum

findepi replied:
> However, this makes other connectors less efficient, e.g. those which don't implement table redirection APIs

isn't it that the difference will be negligible? am i missing something?

findepi replied:
@lxynov was this covered?

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -544,6 +555,14 @@ default void dropView(ConnectorSession session, SchemaTableName viewName)
+     */
+    default Map<SchemaTableName, Optional<ConnectorViewDefinition>> getViewsWithRedirection(ConnectorSession session, Optional<String> schemaName)
+    {
+        return getViews(session, schemaName).entrySet().stream().collect(toUnmodifiableMap(Map.Entry::getKey, entry -> Optional.of(entry.getValue())));
```
Shouldn't the default implementation actually call `redirectTable`?

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -716,6 +735,16 @@ default void revokeTablePrivileges(ConnectorSession session, SchemaTableName tab
+    {
+        return listTablePrivileges(session, prefix).stream()
+                .collect(groupingBy(GrantInfo::getSchemaTableName, toUnmodifiableList())).entrySet().stream()
+                .collect(toUnmodifiableMap(Map.Entry::getKey, entry -> Optional.of(entry.getValue())));
```
Shouldn't the default implementation actually call `redirectTable`?

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -1021,4 +1050,54 @@ default MaterializedViewFreshness getMaterializedViewFreshness(ConnectorSession
+     * <p>
+     * Here are some caveats worth noting:
+     * <li>
+     *     If you choose to implement this method, you should consider implementing {@link #listTableColumnsWithRedirection}, {@link #getViewsWithRedirection},
```
I think the indent here is wrong. At least my IntelliJ removes it upon "Reformat Code" action

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -1021,4 +1050,54 @@ default MaterializedViewFreshness getMaterializedViewFreshness(ConnectorSession
+     * Redirects the target table or view name of a rename operation to another one which may or may not in the same catalog.
+     * <p>
+     * This method is called before the engine tries to rename a table or view. Connectors are expected to return  {@link Optional#empty()} if
+     * the redirection doesn't happen. The engine calls this method together with {@link #redirectTable(ConnectorSession, SchemaTableName)} in
```
> together with ...

It's not clear to me whether `redirectTable(sourceTableName)` is called prior to `redirectTableRename` being called

lxynov replied:
`redirectTableRename` is called with the original source table name, instead of the redirected source table name. So the order doesn't really matter.

findepi replied:
I concluded contrary from the javadoc. Maybe that's just me.

lxynov replied:
It is indeed a little bit confusing.. I've updated the comment

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -1021,4 +1050,54 @@ default MaterializedViewFreshness getMaterializedViewFreshness(ConnectorSession
+     * <li>
+     *     If you choose to implement this method, you should consider implementing {@link #listTableColumnsWithRedirection}, {@link #getViewsWithRedirection},
+     *     {@link #listTablePrivilegesWithRedirection}, too. These methods are called to answer information schema queries. You should also
+     *     consider implementing {@link #redirectTableRename}, which is used to redirect rename targets.
```
what happens if i implement `redirectTable` but do not implement `redirectTableRename`?

lxynov replied:
It really depends. In the Hive-Iceberg case, implementing `redirectTable` and not `redirectTableRename` will cause renames across catalogs, which will fail. However, it can be fine for other cases.

findepi replied:
across catalogs or across schemas?
(renames across catalogs are not possible anyway, right?)

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -1021,4 +1050,54 @@ default MaterializedViewFreshness getMaterializedViewFreshness(ConnectorSession
+     * </li>
+     * <li>
+     *     Be careful if you implement redirection to another catalog. Different catalogs have different sets of roles. In other words, GRANT/REVOKE
+     *     queries may break due to the non-existence of certain roles in the new catalog. Different connectors may have different syntax, feature sets,
```
> syntax

Connectors do not provide syntax. What is this referring to?

lxynov replied:
I was referring to things like
 - table properties (e.g., `partitioned_by` in Hive and `partitioning` in Iceberg)
 - system tables
 - hidden columns

Let me rephrase the "syntax" into details.

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -1021,4 +1050,54 @@ default MaterializedViewFreshness getMaterializedViewFreshness(ConnectorSession
+     * <li>
+     *     Be careful if you implement redirection to another catalog. Different catalogs have different sets of roles. In other words, GRANT/REVOKE
+     *     queries may break due to the non-existence of certain roles in the new catalog. Different connectors may have different syntax, feature sets,
+     *     type support, etc., too. So some working queries with one connector may break with another connector. You should make sure your production
+     *     queries are not among them.
```
That's odd to mention "production queries" here.

This comment is for SPI implementor. From SPI perspective there is no concept of "production queries" or "production" at all. From SPI perspective, breaking any queries is equally bad, not just those regarded as "production".

Actually, SPI implementor and a person responsible for a production environment are not necessarily the same person, so "you" reference here is ambiguous.

Please remove this part.

Actually, all those considerations should be part of `hive.redirect-to-iceberg-enabled` documentation.

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -1021,4 +1050,54 @@ default MaterializedViewFreshness getMaterializedViewFreshness(ConnectorSession
+     *     queries are not among them.
+     * </li>
+     * <li>
+     *     Whether a table is a view or not is determined by the canonical table. E.g., suppose table a.b.c is redirected to x.y.z, it's a view if and only
```
> canonical table

this is not defined and used only here.

if we do not want to define this term, maybe this could say "effective redirect or redirect chain's target"?

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -1021,4 +1050,54 @@ default MaterializedViewFreshness getMaterializedViewFreshness(ConnectorSession
+     * does happen, the engine will access the redirected table or view name instead of the original one and raise a REDIRECTED_TABLE warning.
+     * </p>
+     * <p>
+     * Here are some caveats worth noting:
```
> worth noting

Obviously those _not_ worth noting are not noted here :) 

```suggestion
     * <em>Caveats</em>:
```

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -1021,4 +1050,54 @@ default MaterializedViewFreshness getMaterializedViewFreshness(ConnectorSession
+     * raise a REDIRECTED_TABLE warning.
+     * </p>
+     * <p>
+     * See {@link #redirectTable}'s doc for some note-worthy caveats.
```
```suggestion
     * <em>Caveats</em>: See {@link #redirectTable}'s documentation for caveats of implementing table redirection.
```

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -1021,4 +1050,54 @@ default MaterializedViewFreshness getMaterializedViewFreshness(ConnectorSession
+     *     Whether a table is a view or not is determined by the canonical table. E.g., suppose table a.b.c is redirected to x.y.z, it's a view if and only
+     *     if x.y.z is a view.
+     * </li>
+     * </p>
```
remove (https://stackoverflow.com/questions/5681481/should-ol-ul-be-inside-p-or-outside)
```suggestion
```

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -1021,4 +1050,54 @@ default MaterializedViewFreshness getMaterializedViewFreshness(ConnectorSession
+     *     if x.y.z is a view.
+     * </li>
+     * </p>
+     */
```
It should be explicitly documented here how `redirectTable` plays with access control checks:

- are  control checks applied before redirection takes place
- are  control checks applied after redirection takes place
- are  control checks applied before and after the redirection 

?

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveConfig.java:
```diff
@@ -981,4 +984,31 @@ public HiveConfig setTimestampPrecision(HiveTimestampPrecision timestampPrecisio
+        return this;
+    }
+
+    @NotNull
```
also `@NotEmpty` i guess?

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -636,6 +650,100 @@ private ConnectorTableMetadata doGetTableMetadata(ConnectorSession session, Sche
         return new ConnectorTableMetadata(tableName, columns.build(), properties.build(), comment);
     }
 
+    private ConnectorTableMetadata doGetTableMetadata(ConnectorSession session, SchemaTableName tableName, Table table)
```
This is new method unused. Did you mean to change the existing method instead?

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -223,6 +226,14 @@ default ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTabl
+    /**
+     * Returns a (table name, Optional.empty()) entry if the table is redirected. See {@link #redirectTable}'s doc for details.
+     */
+    default Map<SchemaTableName, Optional<List<ColumnMetadata>>> listTableColumnsWithRedirection(ConnectorSession session, SchemaTablePrefix prefix)
```
If i implement this method, do i still need to implement `listTableColumns`?

same for other `WithRedirection` methods

lxynov replied:
Nope. I'll add to the doc

findepi replied:
Now i understand (https://github.com/prestosql/presto/pull/5160#discussion_r517061612).

Thus it's valid to implement `listTableColumns`, but the method should not be called anymore.
I wish there was a way to express that in Java, but there doesn't seem to be.

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -700,13 +808,18 @@ private ConnectorTableMetadata doGetTableMetadata(ConnectorSession session, Sche
         for (SchemaTableName tableName : listTables(session, prefix)) {
             try {
-                columns.put(tableName, getTableMetadata(session, tableName).getColumns());
+                if (isRedirectToIcebergEnabled(session) && isExistingIcebergTable(session, tableName)) {
```
```
if (redirectTable(...).isPresent())
```

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -716,6 +735,16 @@ default void revokeTablePrivileges(ConnectorSession session, SchemaTableName tab
     }
 
+    /**
+     * Returns a (table name, Optional.empty()) entry if the table is redirected. See {@link #redirectTable}'s doc for details.
+     */
+    default Map<SchemaTableName, Optional<List<GrantInfo>>> listTablePrivilegesWithRedirection(ConnectorSession session, SchemaTablePrefix prefix)
```
What's the point in returning map entries with Optional.empty?
Is _not returning_ them equally correct?

the existing code doesn't seem to use those empty entries, so if there is a future intent, it requires documenting (eg via `@apiNote`)

lxynov replied:
`Optional.empty` means the table is redirected. Since it may be in another catalog, `MetadataManager::listTablePrivileges` should redirect it firstly and then call `ConnectorMetadata::listTablePrivileges`

findepi replied:
> Optional.empty means the table is redirected.

This i understand. But is it for? I didn't find any usage of those empty values, they are just filtered out . Should i search again?

lxynov replied:
Hmm so in `MetadataManager::listTablePrivileges`:
```java
    public List<GrantInfo> listTablePrivileges(Session session, QualifiedTablePrefix prefix)
    {
        ImmutableList.Builder<GrantInfo> grantInfos = ImmutableList.builder();
        ImmutableList.Builder<QualifiedObjectName> redirectedTables = ImmutableList.builder();
        listTablePrivilegesWithRedirection(session, prefix).forEach((name, grantInfo) -> {
            if (grantInfo.isPresent()) {
                grantInfos.addAll(grantInfo.get());
            }
            else {
                redirectedTables.add(new QualifiedObjectName(
                        prefix.getCatalogName(),
                        name.getSchemaName(),
                        name.getTableName()));
            }
        });
        for (QualifiedObjectName redirectedTable : redirectedTables.build()) {
            QualifiedObjectName finalTable = redirectTable(session, redirectedTable, WarningCollector.NOOP);
            listTablePrivilegesWithRedirection(session, finalTable.asQualifiedTablePrefix()).values()
                    .forEach(tableGrantInfos -> tableGrantInfos.ifPresent(grantInfos::addAll));
        }
        return grantInfos.build();
    }
```
When the value is present, grant infos are added to `grantInfos` directly; otherwise the key is put into `redirectedTables` list. And for redirected tables, `listTablePrivilegesWithRedirection` is called again.

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -2465,9 +2578,29 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
+    public Map<SchemaTableName, Optional<List<GrantInfo>>> listTablePrivilegesWithRedirection(ConnectorSession session, SchemaTablePrefix schemaTablePrefix)
     {
-        return accessControlMetadata.listTablePrivileges(session, listTables(session, schemaTablePrefix));
+        Map<SchemaTableName, Optional<ImmutableList.Builder<GrantInfo>>> tablePrivileges = new HashMap<>();
```
Do not use `Map<... ImmutableList.Builder>`, use `Multimap` instead.

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -2465,9 +2578,29 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
+            tablePrivileges.putIfAbsent(tableName, Optional.of(ImmutableList.builder()));
+            tablePrivileges.get(tableName).get().add(grantInfo);
+        });
+        return tablePrivileges.entrySet().stream().collect(toImmutableMap(Entry::getKey, entry -> entry.getValue().isPresent() ? Optional.of(entry.getValue().get().build()) : Optional.empty()));
```
It seems the implementation of `listTablePrivilegesWithRedirection` method could be simplified.
Before suggesting that i have a question on the `listTablePrivilegesWithRedirection` return type. Added there.

findepi commented on presto-hive/src/test/java/io/prestosql/plugin/hive/AbstractTestHive.java:
```diff
@@ -2624,15 +2624,15 @@ public void testHideDeltaLakeTables()
 
                 // list all columns in a schema
-                assertThat(metadata.listTableColumns(session, new SchemaTablePrefix(tableName.getSchemaName())).keySet())
+                assertThat(metadata.listTableColumnsWithRedirection(session, new SchemaTablePrefix(tableName.getSchemaName())).keySet())
```
Why the change here?
Is `listTableColumns` deprecated and should not be called?

Or, we should test both: `listTableColumns` and `listTableColumnsWithRedirection` here?

(this comment applies to other test places as well)

lxynov replied:
`listTableColumns` is removed from `HiveMetadata`

findepi reviewed (dismissed):
See previous comments.

lxynov commented on presto-main/src/main/java/io/prestosql/metadata/Metadata.java:
```diff
@@ -571,4 +573,62 @@ default ResolvedFunction getCoercion(Type fromType, Type toType)
+        }
+
+        @Override
+        public boolean equals(Object obj)
```
There is a `HashSet<RenameRedirection>` object declared in `MetadataManager::redirectTableRename`

lxynov commented:
@electrum @findepi Thanks for the reviews! I've updated the PR. The failing check is unrelated. We can look into cancelled checks later. @findepi I changed the info schema related SPI methods to return `Stream`s as @electrum suggested in https://github.com/prestosql/presto/pull/5160#issuecomment-702337998. We can add a patch later to make `InformationSchemaPageSource` utilize these `Stream` SPIs.

findepi commented:
CI failed due to https://github.com/prestosql/presto/issues/5892 but not only

findepi commented on presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java:
```diff
@@ -354,7 +355,9 @@ public static boolean isTablesEnumeratingTable(InformationSchemaTable table)
                             .map(table -> new QualifiedObjectName(catalogName, prefix.getSchemaName().get(), table)))
-                    .filter(objectName -> !isColumnsEnumeratingTable(informationSchemaTable) || metadata.getTableHandle(session, objectName).isPresent() || metadata.getView(session, objectName).isPresent())
+                    .filter(objectName -> !isColumnsEnumeratingTable(informationSchemaTable) ||
+                            metadata.getTableHandle(session, metadata.redirectTable(session, objectName, WarningCollector.NOOP)).isPresent() ||
+                            metadata.getView(session, metadata.redirectTable(session, objectName, WarningCollector.NOOP)).isPresent())
```
Add a TODO to capture and propagate these warnings.

findepi commented on presto-main/src/main/java/io/prestosql/execution/CreateTableTask.java:
```diff
@@ -96,16 +96,16 @@ public String explain(CreateTable statement, List<Expression> parameters)
 
         Map<NodeRef<Parameter>, Expression> parameterLookup = parameterExtractor(statement, parameters);
-        QualifiedObjectName tableName = createQualifiedObjectName(session, statement, statement.getName());
+        QualifiedObjectName tableName = metadata.redirectTable(session, createQualifiedObjectName(session, statement, statement.getName()), warningCollector);
```
I have a general-level concern that the requirement to call `metadata.redirectTable` before calling a any other method on `metadata` can be error prone in practice, as it seems to easy not to call the `redirectTable` and overlook the missing call also in the review process. I am sure you have thought about this very deeply already... if you happen to have any notes about alternatives considered, i would be curious to learn more.

lxynov replied:
This concern does make sense. I had a similar feeling when I was implementing this PR. However, the table redirection approach seems to be an appropriate one to address the hive-iceberg migration issue. A few other approaches that have been discussed:
 - Federation connector: Introduce a new connector which federates between Hive and Iceberg Connectors.
 - [Search path support](https://github.com/prestosql/presto/issues/4442#issuecomment-658454075)
cc @electrum

findepi replied:
Thanks.
But also if we settle up on redirect, we could try to model the API so that there is a help from type system to make sure redirection is considered. 
Like we have in functions -- function name, resolved function, etc.

findepi commented on presto-main/src/main/java/io/prestosql/metadata/Metadata.java:
```diff
@@ -588,4 +591,62 @@ default ResolvedFunction getCoercion(Type fromType, Type toType)
+        @Override
+        public String toString()
+        {
+            return "(" + sourceTableName + ", " + targetTableName + ")";
```
include "RenameRedirection" (or at least "redirect") in the toString.

Also consider replacing `,` with `=>`

lxynov replied:
Included "Rename" as the pair reflects a rename relationship

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -717,13 +732,15 @@ private ConnectorTableMetadata doGetTableMetadata(ConnectorSession session, Sche
+        return listTables(session, prefix).stream().map(tableName -> {
             try {
-                columns.put(tableName, getTableMetadata(session, tableName).getColumns());
+                if (redirectTable(session, tableName).isPresent()) {
```
Add a check in `getTableHandle` and `getTableHandleForStatisticsCollection` like
```
verify(redirectTable(...).isEmpty(), "Table access should have been redirected for %s", tableName);
```

This will make sure (or improve probability that) if we miss to add `metadata.redirectTable` somewhere (in existing or new code), this will not lead to incorrect results or data corruption.

lxynov replied:
Good idea. Added in them as well as some other methods.

findepi commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -223,6 +224,15 @@ default ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTabl
     }
 
+    /**
+     * Gets the metadata for all columns that match the specified table prefix. Returns a {@link ListTableColumnsResult}
```
The method does not return ListTableColumnsResult. It returns stream of them.

same for other methods

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -737,8 +754,8 @@ private ConnectorTableMetadata doGetTableMetadata(ConnectorSession session, Sche
-        }
-        return columns.build();
+            return null;
+        }).filter(Objects::nonNull);
```
Per general style guidelines, the `.filter` and `.map` above should be preceded with a new line.
I'd recommend to do this as a follow up though, not to incur bigger diff within this PR.

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -2602,9 +2619,26 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
+    public Stream<ListTablePrivilegesResult> listTablePrivilegesStream(ConnectorSession session, SchemaTablePrefix prefix)
     {
-        return accessControlMetadata.listTablePrivileges(session, listTables(session, schemaTablePrefix));
+        if (isRedirectToIcebergEnabled(session)) {
```
Does this  `if` buy as much?

regardless of the case, we call `listTables(session, prefix)`.
then, in redirect case we call `redirectTable` on each item of the list, which is a very cheap operation if redirect is not enabled.
Thus, you could safely handle both cases with single code path.

lxynov replied:
I don't understand clearly... Can you share a code snippet?

findepi replied:
```java
@Override
public Stream<ListTablePrivilegesResult> listTablePrivilegesStream(ConnectorSession session, SchemaTablePrefix prefix)
{
    ImmutableList.Builder<SchemaTableName> redirectedTables = ImmutableList.builder();
    ImmutableList.Builder<SchemaTableName> notRedirectedTables = ImmutableList.builder();
    listTables(session, prefix).forEach(table -> {
        if (redirectTable(session, table).isPresent()) {
            redirectedTables.add(table);
        }
        else {
            notRedirectedTables.add(table);
        }
    });
    return Stream.concat(
            accessControlMetadata.listTablePrivileges(session, notRedirectedTables.build()).stream()
                    .map(grantInfo -> new ListTablePrivilegesResult(grantInfo.getSchemaTableName(), Optional.of(grantInfo))),
            redirectedTables.build().stream()
                    .map(table -> new ListTablePrivilegesResult(table, Optional.empty())));
}
```

findepi replied:
Would that work?

findepi commented on presto-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/environment/singlenode-hive-redirection-to-iceberg/hive.properties:
```diff
@@ -0,0 +1,17 @@
+connector.name=hive-hadoop2
+hive.metastore.uri=thrift://hadoop-master:9083
+hive.config.resources=/docker/presto-product-tests/conf/presto/etc/hive-default-fs-site.xml
+hive.allow-add-column=true
+hive.allow-drop-column=true
+hive.allow-rename-column=true
+hive.allow-comment-table=true
+hive.allow-comment-column=true
+hive.allow-drop-table=true
+hive.allow-rename-table=true
+hive.allow-register-partition-procedure=true
+hive.metastore-cache-ttl=0s
+hive.fs.cache.max-size=10
+hive.max-partitions-per-scan=100
+hive.translate-hive-views=true
```
Are all these needed? i guess not, so please remove them, so the relevant configuration is easier to fish out

lxynov replied:
Removed irrelevant configs.

findepi commented on presto-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/environment/singlenode-hive-redirection-to-iceberg/iceberg.properties:
```diff
@@ -0,0 +1,4 @@
+connector.name=iceberg
+hive.metastore.uri=thrift://hadoop-master:9083
+hive.config.resources=/docker/presto-product-tests/conf/presto/etc/hive-default-fs-site.xml
+iceberg.file-format=PARQUET
```
irrelevant?

lxynov replied:
The Iceberg Connector is needed to test the redirection.

findepi replied:
Surely. I was commenting on the last line of the config only (`iceberg.file-format=PARQUET`)

findepi commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveRedirectionToIceberg.java:
```diff
@@ -0,0 +1,187 @@
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNull;
+
+public class TestHiveRedirectionToIceberg
```
❤️  for a product test!

Did you create or plan to create a QueryRunner-based one too?
That wouldn't make the PT obsolete, but could be helpful for fine-grained testing or do debug things, should anything ever malfunction.

lxynov replied:
Didn't create a QueryRunner-based one because we'll need to install both the Iceberg and Hive plugins in one query runner.

findepi replied:
Can you use both from `presto-tests`?

lxynov commented:
@findepi Thanks for the review! I've updated the PR. @electrum Could you give this another review?

findepi commented on presto-main/src/main/java/io/prestosql/metadata/Metadata.java:
```diff
@@ -588,4 +591,62 @@ default ResolvedFunction getCoercion(Type fromType, Type toType)
+     */
+    RenameRedirection redirectTableRename(Session session, QualifiedObjectName sourceTableName, QualifiedObjectName targetTableName, WarningCollector warningCollector);
+
+    class RenameRedirection
```
make it `final` and perhaps `@Immutable`

phd3 reviewed (commented):
preparatory commits look good. Reviewing main commits.

"Allow multiple warnings with the same warning code"
"Add WarningCollector to data definition task interface"

phd3 reviewed (commented):
Reviewed commit  `Add table redirection methods to SPI`. 

I've the same high level concern  as https://github.com/trinodb/trino/pull/5160#discussion_r521665767. Did we have any further discussions or consensus on it?

sopel39 commented:
@lxynov Could we extract https://github.com/trinodb/trino/pull/5160/commits/76a813fbb791ff99bd7c189072525f545000d329 as a separte PR?

phd3 commented:
@sopel39 I've recently taken over this work from @lxynov, will create a PR with only the SPI changes from #7016 first.

sopel39 commented:
> @sopel39 I've recently taken over this work from @lxynov, will create a PR with only the SPI changes from #7016 first.

Good to know. Thanks!

sopel39 commented:
@phd3 did yo have a chance to extract https://github.com/trinodb/trino/commit/76a813fbb791ff99bd7c189072525f545000d329

phd3 commented:
Superseded by #7606
